CRM/Financial/Form/PaymentEdit.php

```diff
@@ -164,12 +164,48 @@ public function postProcess() {
       $params['check_number'] = CRM_Utils_Array::value('check_number', $this->_submitValues);
     }
 
-    // update the financial trxn
-    civicrm_api3('FinancialTrxn', 'create', $params);
+    if ($this->_submitValues['payment_instrument_id'] != $this->_values['payment_instrument_id']) {
+      //first reverse previous transaction
+      $previousFinanciaTrxn = $this->_values;
+      unset($previousFinanciaTrxn['id'], $params['id']);
+      $previousFinanciaTrxn['trxn_date'] = CRM_Utils_Array::value('trxn_date', $params, date('YmdHis'));
+      $previousFinanciaTrxn['total_amount'] = -$previousFinanciaTrxn['total_amount'];
+      $previousFinanciaTrxn['net_amount'] = -$previousFinanciaTrxn['net_amount'];
+      $previousFinanciaTrxn['fee_amount'] = -$previousFinanciaTrxn['fee_amount'];
+      $previousFinanciaTrxn['to_financial_account_id'] = CRM_Financial_BAO_FinancialTypeAccount::getInstrumentFinancialAccount($params['payment_instrument_id']);
+      $previousFinanciaTrxn['contribution_id'] = self::getContributionIDbyTrxn($this->_id);
+      CRM_Core_BAO_FinancialTrxn::create($previousFinanciaTrxn);
+
+      $params['to_financial_account_id'] = $previousFinanciaTrxn['to_financial_account_id'];
+      $params['contribution_id'] = $previousFinanciaTrxn['contribution_id'];
+      foreach (array('total_amount', 'fee_amount', 'net_amount', 'currency', 'is_payment', 'status_id') as $fieldName) {
+        $params[$fieldName] = $this->_values[$fieldName];
+      }
+      CRM_Core_BAO_FinancialTrxn::create($params);
```

Comment:
So @monishdeb are you saying that if you call the api rather than the BAO here it does something odd? It's inconsistent wit a few rows down